### DISCUSSION TITLE
Useragent to fetch favicons causing bans on websites that use zb-block

### DISCUSSION THREAD

**Discussion by danielb7390** [dani-garcia/vaultwarden#2019] [General]
Hi,
I'm using vaultwarden  for a few months now, loving it so far.

Problem is i keep getting banned on a forum that uses **zb-block** with the rule **IB-UA.999**
The problem seems to be with the `User Agent: Links (2.22; Linux X86_64; GNU C; text)` the rule description is  **auto-bans any UA with 0 spaces in it, OR a space before the 8th character position.**.
So now the question becomes, should vaultwarden use a better user agent than **Links** or it's **zb-block** just too aggressive?

Luckily i know the forum admin and he keeps unbanning me every time this happens, but i wonder how many people got banned in random websites without knowing it.

Thanks in advance.

**Comment by BlackDex** [comment]
First, strange that it blocks, since Links is a well know command line browser.

I added Links, as this, for some reason, works better on cloudflare sites, and some other sites which use JavaScript to check if the browser supports JavaScript or not, while reporting it is chrome or mozilla.

Not sure what your icon timeout is, so, when the refresh happens, since it should just do a few calls to the same site, which in my opinion shouldn't be granted a block.

I had done several user-agent testing, and this one came out to be the better. Not sure if and how to address this the best way.

**Reply by jjlin** [reply]
Links might be well-known in its very small niche, but I wouldn't consider it a well-known browser in general. Might it work better to attempt masquerading as a more common browser first, and then fall back to Links if that fails?

**Reply by BlackDex** [reply]
That could cause a lot of calls if a site doesn't provide a favicon.
It's also hard to be sure if it is caused by user-agent or not in the case there is no actual working html given back, and to just call again using a different user-agent would be a trigger for a ban in my opinion.

An other which we could do, but it's not really something i would like to do, is to add an option to allow a different icon fetching being used instead of the build-in one. Which could be duckduckgo or upstream bitwarden for example.

For me, i think it would be better if people fix that via a reverse proxy to intercept it and redirect instead of Vaultwarden handling all these different cases.

So I'm not clear on how to resolve this in a good way.
Allowing to change the user-agent is also something which can be done, but that is also not something i would prefer to add to the code.

**Reply by jjlin** [reply]
Not sure what you mean by "a lot". It's just one extra call potentially. I think it's more likely that you would get banned for using a (perceived) bogus user agent than for having two requests with different user agents, since that check would be more complex to implement (requiring state to be kept), and it's also very possible for two different clients to legitimately share an IP.

**Reply by BlackDex** [reply]
Lets keep bogus user-agent as a discussion point, since it still is a browser used, also, Vaultwarden tries to fake a browser, and a lot of other protections check if it is valid, so changing user-agent again is not going to solve anything. Yes it will solve this specific case, but it could cause issues with others, that would be going to become a ping-pong debate and switcheroo.

Also, if you would check the code, it does a lot more then just one or two calls if it didn't got a valid response the first time, which is to much to explain here.

**Comment by Maikuolan** [comment]
Hi,

I found this discussion per chance, after performing a vanity search for ZB Block just now (I'm not the developer/maintainer for ZB Block, but I know the current plus former developers/maintainers for ZB Block, have assisted in the past, and I'm active in its community, so I'm familiar with the ins-and-outs of ZB Block, and sometimes get curious about potential new search results, etc).

> Might it work better to attempt masquerading as a more common browser first, and then fall back to Links if that fails?

Please don't do that. I understand that such practice is rather common and widespread among similar such projects, but if there's one thing that most ZB Block users tend to not tolerate, and tend to actively try blocking, it's non-browsers masquerading as browsers, or those things which fake their own identity, in an attempt to pretend to be something else. It should be possible to resolve the problem raised by this discussion easily enough, but masquerading as legitimate browsers is something which might potentially risk raising some hostility from the ZB Block community (which might pose a hindrance if the need to resolve other similar problems in the future arises).

> So now the question becomes, should vaultwarden use a better user agent than **Links** or it's **zb-block** just too aggressive?

Good to make a distinction here: The block events in question are occurring due to a singular, specific signature (`IB-UA.999`). So, it's not a matter of whether ZB Block *as a whole* is "too aggressive", seeing as it's just that one signature causing problems here, not the entirety of ZB Block.

However, having looked at the specific signature in question, I can definitely see the potential for false positives potentially occurring against other projects too, so I do feel personally that the specific signature in question is somewhat too aggressive. The signature in question appears to trigger for any user agents containing a space occurring at the seventh byte or earlier. Seeing as, for the user agent in question in this case, there's a space occurring at the fifth position, the signature is triggering.

I'm not sure offhandedly what inspired that signature in the first place, whether there are any specific known malicious bots, user agents, etc which that signature is intended to catch. In any case, I'll drop a message to the current developer/maintainer to bring their attention to this discussion, in case they want to do anything about the matter.

If the current developer/maintainer decides to remove or modify the signature in question, that would resolve the problem here, at least, for any future ZB Block installations, and for those which choose to update to the latest signatures (after said removal/modification has been made, that is), in which case, you shouldn't need to worry about changing your user agent. Of course, we're talking only future installations and installations which update their signatures. The problem would still persist at those websites with owners which forget to update.

Alternatively, or as a short-term solution, you could probably just add "Vaultwarden" or something like that to the beginning of your user agent to avoid triggering the signature in question, seeing as that would push the space up to the 12th positon at the earliest.
